blockchain/stake/internal/tickettreap/immutable.go

```diff
@@ -351,6 +378,46 @@ func (t *Immutable) ForEach(fn func(k Key, v *Value) bool) {
 	}
 }
 
+// ForEachByHeight iterates all elements in the tree less than a given height in the blockchain
+func (t *Immutable) ForEachByHeight(heightLessThan uint32, fn func(k Key, v *Value) bool) {
+	// Add the root node and all children to the left of it to the list of
+	// nodes to traverse and loop until they, and all of their child nodes,
+	// have been traversed.
+	var parents parentStack
+	for node := t.root; node != nil && node.priority < heightLessThan; node = node.left {
+		parents.Push(node)
+	}
+	for parents.Len() > 0 {
+		node := parents.Pop()
+		if !fn(node.key, node.value) {
+			return
+		}
+
+		// Extend the nodes to traverse by all children to the left of
+		// the current node's right child.
+		for node := node.right; node != nil && node.priority < heightLessThan; node = node.left {
+			parents.Push(node)
+		}
+	}
+}
+
+func (t* Immutable) RootPriority() uint32 {
+	if t.root != nil {
+		return t.root.priority
+	}
+	return 0
+}
+
+// IsHeap validates that the invariant of the min-heap, used for testing
+func (t* Immutable) IsHeap() bool {
```

Comment:
Testing only functions must not be exposed in production code.  Besides, the test code is in the same package, but only compiled when the tests are run, which means it can just access `t.root.isHeap()` directly anyways.